<a href="https://colab.research.google.com/github/byambaa1982/python_cources_for_beginers/blob/grace/titanic_data_analysis/basic_dml.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import sqlite3


# SQL

- Step 1. Download titanic.sqlite

- step 2. Upload file into notebook

- step 3. Run the first two cells

In [2]:
# Connect to the SQLite database
conn = sqlite3.connect('titanic.sqlite')

# Read the 'titanic' table into a pandas DataFrame
df = pd.read_sql('SELECT * FROM titanic', conn)
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500  None        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250  None        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500  None        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000  None        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500  None        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500  None        Q  

[891 rows x 12 columns]

# Use 'PassengerId', 'Name', 'Sex', 'Age' columns and save dataframe into database as passenger_info

In [40]:
# 1. Passenger Info Table
passenger_info = df[['PassengerId', 'Name', 'Sex', 'Age']]
passenger_info.to_sql('passenger_info', conn, if_exists='replace', index=False)

# 2. Ticket Info Table
ticket_info = df[['PassengerId', 'Ticket', 'Fare', 'Pclass']]
ticket_info.to_sql('ticket_info', conn, if_exists='replace', index=False)

# 3. Survival Status Table
survival_status = df[['PassengerId', 'Survived', 'Parch', 'SibSp']]
survival_status.to_sql('survival_status', conn, if_exists='replace', index=False)


891

In [4]:
tables = pd.read_sql("SELECT name FROM sqlite_master WHERE type='table'", conn)
print(tables)

      name
0  titanic


In [7]:
q = '''
select PassengerId, Name, Sex, Age from passenger_info limit 5
'''
p_info = pd.read_sql(q, conn)
p_info

PassengerId                                               Name     Sex  \
0            1                            Braund, Mr. Owen Harris    male   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2            3                             Heikkinen, Miss. Laina  female   
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4            5                           Allen, Mr. William Henry    male   

    Age  
0  22.0  
1  38.0  
2  26.0  
3  35.0  
4  35.0

In [8]:
delete_query = '''
DELETE FROM passenger_info
WHERE PassengerId = 3
'''
cursor = conn.cursor()
cursor.execute(delete_query)
conn.commit()


In [9]:
q = '''
select PassengerId, Name, Sex , Age from passenger_info limit 5
'''
p_info = pd.read_sql(q, conn)
p_info

PassengerId                                               Name     Sex  \
0            1                            Braund, Mr. Owen Harris    male   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
3            5                           Allen, Mr. William Henry    male   
4            6                                   Moran, Mr. James    male   

    Age  
0  22.0  
1  38.0  
2  35.0  
3  35.0  
4   NaN

In [10]:
delete_query = '''
update passenger_info
set Age = 23
WHERE PassengerId = 6
'''
cursor = conn.cursor()
cursor.execute(delete_query)
conn.commit()

In [11]:
q = '''
select PassengerId, Name, Sex , Age from passenger_info limit 5
'''
p_info = pd.read_sql(q, conn)
p_info

PassengerId                                               Name     Sex  \
0            1                            Braund, Mr. Owen Harris    male   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
3            5                           Allen, Mr. William Henry    male   
4            6                                   Moran, Mr. James    male   

    Age  
0  22.0  
1  38.0  
2  35.0  
3  35.0  
4  23.0

In [12]:
insert_query = '''
INSERT INTO
  passenger_info (PassengerId, Name, Sex, Age)
VALUES
  (3, 'Sara Lucas', 'female', 45.5)
'''
cursor = conn.cursor()
cursor.execute(insert_query)
conn.commit()

In [14]:
q = '''
select
  PassengerId, Name, Sex, Age
from
  passenger_info
order by
  PassengerId
limit
  5
'''
p_info = pd.read_sql(q, conn)
p_info

PassengerId                                               Name     Sex  \
0            1                            Braund, Mr. Owen Harris    male   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2            3                                         Sara Lucas  female   
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4            5                           Allen, Mr. William Henry    male   

    Age  
0  22.0  
1  38.0  
2  45.5  
3  35.0  
4  35.0

# Task 1: Update Task
Objective:

- Update the age of a passenger based on their PassengerId.

Instructions:

- Update the age of the passenger with PassengerId = 5 in the passenger_info table to 40.

In [22]:

update_query = '''
update
  passenger_info
set
  Age = 40
where
  PassengerId = 5
'''
cursor = conn.cursor()
cursor.execute(update_query)
conn.commit()

q = '''
select
PassengerId, Name, Sex, Age
from
    passenger_info
limit
  10
'''

p_info = pd.read_sql(q, conn)
p_info

PassengerId                                               Name     Sex  \
0            1                            Braund, Mr. Owen Harris    male   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
3            5                           Allen, Mr. William Henry    male   
4            6                                   Moran, Mr. James    male   
5            7                            McCarthy, Mr. Timothy J    male   
6            8                     Palsson, Master. Gosta Leonard    male   
7            9  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   
8           10                Nasser, Mrs. Nicholas (Adele Achem)  female   
9           11                    Sandstrom, Miss. Marguerite Rut  female   

    Age  
0  22.0  
1  38.0  
2  35.0  
3  40.0  
4  23.0  
5  54.0  
6   2.0  
7  27.0  
8  14.0  
9   4.0

# Task 2: Delete Task
Objective:

- Delete a passenger's ticket information based on their PassengerId.

Instructions:

- Delete the record of the passenger with PassengerId = 8 from the ticket_info table.

In [23]:

delete_query = '''
delete from
  passenger_info
where
  PassengerId = 8
'''
cursor = conn.cursor()
cursor.execute(delete_query)
conn.commit()

q = '''
select
PassengerId, Name, Sex, Age
from
    passenger_info
limit
  10
'''

p_info = pd.read_sql(q, conn)
p_info

PassengerId                                               Name     Sex  \
0            1                            Braund, Mr. Owen Harris    male   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
3            5                           Allen, Mr. William Henry    male   
4            6                                   Moran, Mr. James    male   
5            7                            McCarthy, Mr. Timothy J    male   
6            9  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   
7           10                Nasser, Mrs. Nicholas (Adele Achem)  female   
8           11                    Sandstrom, Miss. Marguerite Rut  female   
9           12                           Bonnell, Miss. Elizabeth  female   

    Age  
0  22.0  
1  38.0  
2  35.0  
3  40.0  
4  23.0  
5  54.0  
6  27.0  
7  14.0  
8   4.0  
9  58.0

# Task 3: Insert Task
Objective:

- Insert a new passenger's survival status into the survival_status table.

Instructions:

- Insert a new record with the following details into the survival_status table:
- PassengerId = 891
- Survived = 1
- Parch = 0
- SibSp = 1

In [36]:
insert_query = '''
insert into
  survival_status (PassengerId, Survived, Parch, SibSp)
values
  (891, 1, 0, 1)
'''
cursor = conn.cursor()
cursor.execute(insert_query)
conn.commit()


select_query = '''
select
  PassengerId, Survived, Parch, SibSp
from
  survival_status
where
  PassengerId = 891
limit
  2
'''

p_info = pd.read_sql(select_query, conn)
p_info

PassengerId  Survived  Parch  SibSp
0          891         0      0      0
1          891         1      0      1

# Task 4: Update Multiple Columns
Objective:

- Update multiple columns for a specific passenger in the ticket_info table.

Instructions:

- Update the Fare to 150.0 and Pclass to 1 for the passenger with PassengerId = 4 in the ticket_info table.

In [42]:
update_query = '''
update
  ticket_info
set
   Fare = 150,
   Pclass = 1
where
  PassengerId = 4
'''
cursor = conn.cursor()
cursor.execute(update_query)
conn.commit()


select_query = '''
select
  PassengerId, Ticket, Fare, Pclass
from
  ticket_info
where
  PassengerId = 4
'''

p_info = pd.read_sql(select_query, conn)
p_info

PassengerId  Ticket   Fare  Pclass
0            4  113803  150.0       1

# Task 5: Delete Based on Condition
Objective:

- Delete records where a condition is met in the survival_status table.

Instructions:

- Delete all passengers from the survival_status table where Survived = 0 (i.e., those who did not survive).

In [45]:
delete_query = '''
delete from
  survival_status
where
  Survived = 0
'''
cursor = conn.cursor()
cursor.execute(delete_query)
conn.commit()


select_query = '''
select
  PassengerId, Survived
from
  survival_status
where
  PassengerId = 891
'''

p_info = pd.read_sql(select_query, conn)
p_info

Empty DataFrame
Columns: [PassengerId, Survived]
Index: []

# Task 6: Insert New Passenger with Complete Information
Objective:

- Insert a new passenger into all three tables: passenger_info, ticket_info, and survival_status.

Instructions:

Insert a new passenger with the following details:
- PassengerId = 892
- Name = 'Tom Hardy', Sex = 'male', Age = 37 into passenger_info
- Ticket = 'PC 17599', Fare = 75.0, Pclass = 1 into ticket_info
- Survived = 1, Parch = 0, SibSp = 0 into survival_status

In [53]:
insert_query = '''
INSERT INTO passenger_info (PassengerId, Name, Sex, Age)
VALUES (892, 'Tom Hardy', 'male', 37)
'''
cursor = conn.cursor()
cursor.execute(insert_query)
conn.commit()

select_query = '''
select
  PassengerId, Name, Sex, Age
from
  passenger_info
where
  PassengerId = 892
limit
  1
'''

p_info = pd.read_sql(select_query, conn)
p_info


PassengerId       Name   Sex   Age
0          892  Tom Hardy  male  37.0

In [ ]:
insert_query = '''
INSERT INTO Ticket (PassengerId, Name, Sex, Age)
VALUES (892, 'Tom Hardy', 'male', 37)
'''
cursor = conn.cursor()
cursor.execute(insert_query)
conn.commit()

select_query = '''
select
  PassengerId, Name, Sex, Age
from
  passenger_info
where
  PassengerId = 892
limit
  1
'''

p_info = pd.read_sql(select_query, conn)
p_info
